plugins/image-prioritizer/class-image-prioritizer-background-image-styled-tag-visitor.php

```diff
@@ -65,33 +80,128 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 		}
 
 		if ( is_null( $background_image_url ) ) {
+			$this->maybe_preload_external_lcp_background_image( $context );
 			return false;
 		}
 
 		$xpath = $processor->get_xpath();
 
 		// If this element is the LCP (for a breakpoint group), add a preload link for it.
 		foreach ( $context->url_metric_group_collection->get_groups_by_lcp_element( $xpath ) as $group ) {
```

Comment:
This is missing the fix we're including via #1723. I think we should merge that PR since it's more of a fix while this is an enhancement, and then update this PR to follow suit. Otherwise we're introducing the same problem for background images that #1723 is fixing for the LCP.

This re-emphasizes to me what I mentioned in my previous point: If we were using Optimization Detective's own LCP handling, we would be able to get this kind of correct behavior "for free", rather than having to re-implement it here too. So I'm curious why we're not doing that.

At a minimum, this needs to be fixed here to be aligned with what #1723 does.

Reply:
> This is missing the fix we're including via #1723.

I'm not sure this is the case. The fix in #1723 is to allow `fetchpriority=high` to be on the `IMG` element when only the mobile and desktop viewport groups have URL Metrics collected. Preload links are still only being added for `IMG` for the actual viewport groups that have URL metrics collected for them:

https://github.com/WordPress/performance/blob/d13f33c8f561f6d61fe8e5d20c2a2192904af493/plugins/image-prioritizer/class-image-prioritizer-img-tag-visitor.php#L342-L348

Or are you saying that if the mobile and desktop viewport groups have the same LCP element, that we should go ahead and also preload the same URL for that LCP Element even for phablet and tablet when no URL Metrics are yet confirming that they also have the same LCP element?

> This re-emphasizes to me what I mentioned in my previous point: If we were using Optimization Detective's own LCP handling, we would be able to get this kind of correct behavior "for free", rather than having to re-implement it here too. So I'm curious why we're not doing that.

I'm not sure I understand. For elements which are tracked in URL Metrics (and an element appears in `elements` when there is a tag visitor which returns `true` for a tag during iterator), then as part of the stored metadata for that tracked element will be whether it `isLCP`. So here it's looking up which viewport groups have the current tag as the LCP element, and for each one a preload link is added.